### PR TITLE
Theme.json: Remove custom prefixes from properties that did not land in 5.8

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related: https://github.com/WordPress/gutenberg/pull/31366#issuecomment-907075652, https://github.com/WordPress/gutenberg/pull/31366#issuecomment-908135526

## Description

This PR removes the `custom` prefix from properties within theme.json that do not require it. Those that remain have already landed in WP 5.8 and will be addressed in the future by adding a migration from `v1` to the future `v2` theme.json schema.

## How has this been tested?

- Unit tests
- Inspecting settings via the store in block and site editors
- Manually testing various combinations of prefixed and unprefixed flags
- Switching between themes using old and new flags

#### Testing instructions

1. Run the following tests:
```
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php
```
```
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-schema-v0-test.php
```
```
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-schema-v1-test.php
```
	

2. Setup your theme.json to turn all the border and typography supports in the snippet below to `false`. 
    [Example gist here](https://gist.github.com/aaronrobertshaw/4934e8f308d1d747fef6ee7726440137)
```json
{
	"settings": {
		"border": {
			"color": false,
			"customRadius": false,
			"style": false,
			"width": false
		},
		"typography": {
			"fontStyle": false,
			"fontWeight": false,
			"textDecoration": false,
			"textTransform": false,
			"letterSpacing": false,
		},
	}
}
```
3. Create a new post and add a group block and navigation block to the post and save
4. Select the group block and ensure the border panel and its controls are not present
5. Confirm toggling the individual border supports on in your theme.json is reflected in the available controls when selecting the group block
6. Select the navigation block and ensure that none of the following controls show in the typography panel:
	* `fontStyle`
	* `fontWeight`
	* `textDecorations`
	* `textTransforms`
	* `letterSpacing`
7. Toggle each of the above properties on in your theme.json and confirm their related controls now display in the navigation block’s typography panel.
8. Switch to the site editor and ensure the correct controls also display in the relevant block panels there.
	* Note: Text decoration and text-transform have not been exposed within global styles to date. 
9. Change individual support flags back to their `custom` prefixed form e.g. `customFontStyle` or `customColor`. Then confirm the relevant controls still display correctly in both editors.
10. Test that `useSetting` will return the correct value when using the deprecated `custom` prefixed flag.
<details>
  <summary>Example patch to test `useSetting` with deprecated flags</summary>
  
```diff
diff --git a/packages/block-editor/src/hooks/border.js b/packages/block-editor/src/hooks/border.js
index 761c04200a..37a5213c27 100644
--- a/packages/block-editor/src/hooks/border.js
+++ b/packages/block-editor/src/hooks/border.js
@@ -23,17 +23,17 @@ export function BorderPanel( props ) {
 	const isSupported = hasBorderSupport( props.name );
 
 	const isColorSupported =
-		useSetting( 'border.color' ) && hasBorderSupport( props.name, 'color' );
+		useSetting( 'border.customColor' ) && hasBorderSupport( props.name, 'color' );
 
 	const isRadiusSupported =
-		useSetting( 'border.radius' ) &&
+		useSetting( 'border.customRadius' ) &&
 		hasBorderSupport( props.name, 'radius' );
 
 	const isStyleSupported =
-		useSetting( 'border.style' ) && hasBorderSupport( props.name, 'style' );
+		useSetting( 'border.customStyle' ) && hasBorderSupport( props.name, 'style' );
 
 	const isWidthSupported =
-		useSetting( 'border.width' ) && hasBorderSupport( props.name, 'width' );
+		useSetting( 'border.customWidth' ) && hasBorderSupport( props.name, 'width' );
 
 	if ( isDisabled || ! isSupported ) {
 		return null;
@@ -110,10 +110,10 @@ export function shouldSkipSerialization( blockType ) {
  */
 const useIsBorderDisabled = () => {
 	const configs = [
-		! useSetting( 'border.color' ),
-		! useSetting( 'border.radius' ),
-		! useSetting( 'border.style' ),
-		! useSetting( 'border.width' ),
+		! useSetting( 'border.customColor' ),
+		! useSetting( 'border.customRadius' ),
+		! useSetting( 'border.customStyle' ),
+		! useSetting( 'border.customWidth' ),
 	];
 
 	return configs.every( Boolean );

```
</details>

## Types of changes
Enhancement.
Breaking change ( 3rd parties working directly with theme.json or store shape will need updating )

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by oandregal** [comment]
I'm thinking about what @mtias mentioned [here](https://github.com/WordPress/gutenberg/issues/34349#issuecomment-907111110). Perhaps we can build two back-compatibility bridges to cover most of the cases and give people time to adapt? I'm thinking of:

- Both `useSetting( 'border.customRadius' )` and `useSetting( 'border.radius' )` will work. This can be built within `useSetting`.
- Both `settings.border.customRadius` and `settings.border.radius` will work in `theme.json`. This can be built within the WP_Theme_JSON constructor with a `WP_Theme_JSON_Schema_V1::parse( $ settings )` in the same way we have for v0.

These bridges can be in place for a few Gutenberg versions or until we land this in WordPress 5.9. We should give a heads up in the Gutenberg release post or adding deprecations, etc.

I reckon there're still ways in which this change forces 3rd parties to adapt: for example, if they're working directly with the `theme.json` object shape or if they work directly with the `core/block-editor` store shape instead of using the tools available. I think this is an ok trade-off if we want to migrate these names as per https://github.com/WordPress/gutenberg/issues/34349

cc @jorgefilipecosta for thoughts.

**Comment by aaronrobertshaw** [comment]
> I'm thinking about what mtias mentioned here.

I missed that comment, thanks.

> Perhaps we can build two back-compatibility bridges to cover most of the cases and give people time to adapt? 

Your suggestions work for me. I'm happy to put this PR on ice and close it after splitting out the minor bug fixes.

**Comment by aaronrobertshaw** [comment]
Closing this PR as the minor bug fixes are in their own PRs now.

- https://github.com/WordPress/gutenberg/pull/34515
- https://github.com/WordPress/gutenberg/pull/34516

**Comment by oandregal** [comment]
Nice work on splitting the PR to land faster the bugfixes :clap: 

My comment wasn't meant to close this but to add those two things! My stance would be that those are enough given the experimental nature of the existing properties. It's also best to do it as far as possible from the integration cycle (we want to give people plenty of time to adapt instead of having to change things at the last minute).

**Comment by aaronrobertshaw** [comment]
Ok, thanks, I jumped the gun closing this one then.

I'll add the requested updates to useSetting along with a new theme.json schema class to map the custom prefixed properties over to their newer versions for v1.

**Comment by aaronrobertshaw** [comment]
I've rebased this PR to bring it up-to-date.

The `useSetting` hook has been updated to support both the prefixed and nonprefixed forms of theme.json flags that didn't land in 5.8.

A new v1 schema class was added as an interim measure to migrate the prefixed flags to their nonprefixed version within the `WP_Theme_JSON` constructor as [suggested](https://github.com/WordPress/gutenberg/pull/34485#issuecomment-911333007). A full migration of the current v1 schema to a new v2 schema, will occur in the future where the remaining properties with a `custom` prefix are also updated.

It's entirely possible I've missed something here so I'll give this further testing tomorrow. In the meantime, @oandregal can you confirm we still want these changes to land for 5.9 and the updates to `useSetting` etc were in line with your expectations?

File: docs/how-to-guides/themes/theme-json.md [hunk 1]
```diff
@@ -216,10 +216,10 @@ The settings section has the following structure:
 	"version": 1,
 	"settings": {
 		"border": {
-			"customColor": false,
-			"customRadius": false,
-			"customStyle": false,
-			"customWidth": false
+			"color": false,
+			"radius": false,
```
Comment (oandregal) [inline docs/how-to-guides/themes/theme-json.md]: Apparently, `border.customRadius` landed in WordPress 5.8 ([see](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/class-wp-theme-json.php#L181)), so it'd be best to leave this for a separate PR where we migrate the existing to the v2.

File: lib/class-wp-theme-json-gutenberg.php [hunk 2]
```diff
@@ -82,10 +82,10 @@ class WP_Theme_JSON_Gutenberg {
 
 	const VALID_SETTINGS = array(
 		'border'     => array(
-			'customColor'  => null,
-			'customRadius' => null,
-			'customStyle'  => null,
-			'customWidth'  => null,
+			'color'  => null,
+			'radius' => null,
```
Comment (oandregal) [inline lib/class-wp-theme-json-gutenberg.php]: Apparently, border.customRadius landed in WordPress 5.8 ([see](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/class-wp-theme-json.php#L181)), so it'd be best to leave this for a separate PR where we migrate the existing to the v2.

File: docs/how-to-guides/themes/create-block-theme.md [hunk 3]
```diff
@@ -439,10 +439,10 @@ To enable border styles, add a `border` object under `settings` with the followi
 	"version": 1,
 	"settings": {
 		"border": {
-			"customColor": true,
-			"customRadius": true,
-			"customStyle": true,
-			"customWidth": true
+			"color": true,
+			"radius": true,
```
Comment (oandregal) [inline docs/how-to-guides/themes/create-block-theme.md]: Same here about `border.customRadius`.

File: lib/class-wp-theme-json-gutenberg.php (hunk 2 above)
Comment (aaronrobertshaw) [reply lib/class-wp-theme-json-gutenberg.php]: Thanks for that 👍 

I'm a bit slow today. To clarify, you are only suggesting that we exclude `border.customRadius` from this PR not leave all `custom` prefix removal until we move to a v2, correct?

Comment (oandregal) [reply lib/class-wp-theme-json-gutenberg.php]: Exactly, only exclude radius from here.

Comment (aaronrobertshaw) [reply lib/class-wp-theme-json-gutenberg.php]: No worries. I'll fix that up first thing tomorrow and double-check the other flags.

File: packages/block-editor/src/components/use-setting/index.js [hunk 4]
```diff
@@ -72,10 +103,12 @@ export default function useSetting( path ) {
 
 			// 1 - Use __experimental features, if available.
 			// We cascade to the all value if the block one is not available.
-			const defaultsPath = `__experimentalFeatures.${ path }`;
```
Comment (oandregal) [inline packages/block-editor/src/components/use-setting/index.js:75]: Instead of adding a new step, what would you think of pre-processing the given path? Similar to what we do with `WP_Theme_JSON_Schema_V1::parse()`. E.g.:

```js
const path = removeCustomPrefixes( path );

// Follow existing code.
```

File: lib/class-wp-theme-json-gutenberg.php (hunk 2 above)
Comment (aaronrobertshaw) [reply lib/class-wp-theme-json-gutenberg.php]: The `custom` prefix has been reinstated for the `border.customRadius` flag: https://github.com/WordPress/gutenberg/pull/34485/commits/2e7149430dee7d6e6c84f4270a41b810ae95e82a

File: packages/block-editor/src/components/use-setting/index.js (hunk 4 above)
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/use-setting/index.js:75]: Much better. I've implemented that change. Thank you 👍

**Review by oandregal — APPROVED** [review]
Nice, thank you!
